File: lib/models/download_task.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'media_item.dart';

part 'download_task.freezed.dart';
part 'download_task.g.dart';

enum DownloadState { queued, running, paused, completed, failed, canceled }

@freezed
abstract class DownloadTask with _$DownloadTask {
  const factory DownloadTask({
    required String id,
    required MediaItem mediaItem,
    @Default(0) double progress,
    @Default(DownloadState.queued) DownloadState state,
    String? savedPath,
    String? errorMessage,
  }) = _DownloadTask;

  factory DownloadTask.fromJson(Map<String, dynamic> json) =>
      _$DownloadTaskFromJson(json);
}

File: app/src/main/java/io/github/rustnomicon/sourceinstaller/data/model/DownloadTask.kt
package io.github.rustnomicon.sourceinstaller.data.model

enum class DownloadState { Queued, Running, Paused, Completed, Failed, Canceled }

/**
 * Port of the Flutter `DownloadTask` freezed model.
 */
data class DownloadTask(
    val id: String,
    val mediaItem: MediaItem,
    val progress: Float = 0f,
    val state: DownloadState = DownloadState.Queued,
    val savedPath: String? = null,
    val errorMessage: String? = null,
)
